tenders/spt/spt_elf.c

```diff
@@ -0,0 +1,227 @@
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * spt_elf.c: ELF loader.
+ *
+ * This module should be kept backend-independent and architectural
+ * dependencies should be self-contained.
+ */
+
+#define _GNU_SOURCE
+#include <err.h>
+#include <elf.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "spt.h"
+
+static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
+{
+    ssize_t total = 0;
+    char *p = buf;
+
+    if (count > SSIZE_MAX) {
+        errno = E2BIG;
+        return -1;
+    }
+
+    while (count > 0) {
+        ssize_t nr;
+
+        nr = pread(fd, p, count, offset);
+        if (nr == 0)
+            return total;
+        else if (nr == -1 && errno == EINTR)
+            continue;
+        else if (nr == -1)
+            return -1;
+
+        count -= nr;
+        total += nr;
+        p += nr;
+        offset += nr;
+    }
+
+    return total;
+}
+
+/*
+ * Load code from elf file into *mem and return the elf entry point
+ * and the last byte of the program when loaded into memory. This
+ * accounts not only for the last loaded piece of code from the elf,
+ * but also for the zeroed out pieces that are not loaded and sould be
+ * reserved.
+ *
+ * Memory will look like this after the elf is loaded:
+ *
+ * *mem                    *p_entry                   *p_end
+ *   |             |                    |                |
+ *   |    ...      | .text .rodata      |   .data .bss   |
+ *   |             |        code        |   00000000000  |
+ *   |             |  [PROT_EXEC|READ]  |                |
+ *
+ */
+void spt_elf_load(const char *file, uint8_t *mem, size_t mem_size,
+       uint64_t *p_entry, uint64_t *p_end)
+{
+    int fd_kernel;
+    ssize_t numb;
+    size_t buflen;
+    Elf64_Off ph_off;
+    Elf64_Half ph_entsz;
+    Elf64_Half ph_cnt;
+    Elf64_Half ph_i;
+    Elf64_Phdr *phdr = NULL;
+    Elf64_Ehdr hdr;
+
+    /* elf entry point (on physical memory) */
+    *p_entry = 0;
+    /* highest byte of the program (on physical memory) */
+    *p_end = 0;
+
+    fd_kernel = open(file, O_RDONLY);
```

Comment:
`fd_kernel` will be closed only if we did not go to `out_error` or `out_invalid` which is fine for me - because `err` will stop the program - but it's a file-descriptor leak.

Reply:
Thanks, this is minor, can you please file an issue so that I can fix it in both `spt` and `hvt` after this is merged.

Reply:
Done by #311